### PR TITLE
#25 transform

### LINKED ISSUE #25: Transform meshes
Apply simple transforms, to meshes.
Blocked by #24

### PR DESCRIPTION
**Pull request by Sckorn**
Fix #25

### CODE REVIEW

**Comment by radameon** [comment]
Add Transform related test, until then, review is redundant.

**Comment by radameon** [comment]
Also, implement quaternion conjunction and inverse.

**Comment by Sckorn** [comment]
Transform tests introduced. Quaternion fully implemented.

File: engine_datamodel/include/data_types/math/Matrix.hpp [hunk 1]
```diff
@@ -32,30 +92,95 @@ class Matrix
         std::copy(m.begin(), m.end(), _m.begin());
     }
 
-    ~Matrix()
+    template <typename InputIterator>
+    Matrix(InputIterator begin, InputIterator end)
     {
-        static_assert(std::is_scalar_v<ValueType>);
-    };
+        assert(begin != end);
```
Comment (radameon) [inline engine_datamodel/include/data_types/math/Matrix.hpp:98]: Replace asserts with exception throws

File: engine_datamodel/include/data_types/math/Matrix.hpp [hunk 2]
```diff
@@ -32,30 +92,95 @@ class Matrix
...
+        return column_type {_tmp};
+    }
+
+    // row-major
+    const value_type &at(size_t row, size_t col) const
+    {
+        return _m[row * DimRow + col];
+    }
+
+    Matrix transpose()
```
Comment (radameon) [inline engine_datamodel/include/data_types/math/Matrix.hpp:135]: Mark this const

File: engine_datamodel/include/data_types/math/Quaternion.hpp [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+        auto scalar = scal1 * scal2 - vec1 * vec2;
+        auto vec = scal1 * vec2 + scal2 * vec1 + Quaternion::vector_type::cross(vec1, vec2);
+
+        return {vec, scalar};
+    }
+
+    Quaternion conjugate() const 
+    {
+        auto scalar = w();
+        auto vec = rcbe::math::Vector3d {x(), y(), z()} * -1;
```
Comment (radameon) [inline engine_datamodel/include/data_types/math/Quaternion.hpp:152]: Use `getVectorPart`

File: engine_datamodel/include/data_types/math/Quaternion.hpp [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+    Quaternion inverse() const
+    {
+        auto norm_val = norm();
+        norm_val *= norm_val;
+        norm_val = 1 / norm_val;
+
+        auto conj = conjugate();
+
+        auto scalar = conj.w() * norm_val;
+        vector_type tmp {conj.x(), conj.y(), conj.z()};
```
Comment (radameon) [inline engine_datamodel/include/data_types/math/Quaternion.hpp:165]: Use `getVectorPart`

File: engine_datamodel/include/data_types/math/Transform.hpp [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    const rotation_type &getRotation() const;
+    rotation_type &getRotation();
+
+    const translation_type &getTranslation() const;
+    translation_type &getTranslation();
+
+    Vector3d transform(const Vector3d &v) const;
+    Vector3d operator()(const Vector3d &v) const;
+    friend Vector3d operator*(const Transform &t, const Vector3d &v);
+    friend Transform operator*(const Transform &lhs, const Transform &rhs);
```
Comment (radameon) [inline engine_datamodel/include/data_types/math/Transform.hpp:27]: Add `operator*(const Transform&, const Matrix3x3)`

File: engine_datamodel/src/geometry/Mesh.cpp [hunk 6]
```diff
@@ -96,6 +96,28 @@ const Mesh::color_type &Mesh::color() const
 {
     return _color;
 }
+
+const Mesh::transform_type Mesh::transform() const
```
Comment (radameon) [inline engine_datamodel/src/geometry/Mesh.cpp:100]: See comments in `hpp` file.

File: engine_datamodel/src/geometry/Mesh.cpp [hunk 7]
```diff
@@ -96,6 +96,28 @@ const Mesh::color_type &Mesh::color() const
...
+{
+    return _transform;
+}
+
+void Mesh::set_transform(const transform_type &t)
```
Comment (radameon) [inline engine_datamodel/src/geometry/Mesh.cpp:105]: See comments in `hpp` file.

File: engine_datamodel/src/math/Matrix.cpp [hunk 8]
```diff
@@ -2,22 +2,120 @@
...
-void adl_serializer<rcbe::math::Matrix<rcbe::core::EngineScalar, 3, 3>> ::from_json(
-    const nlohmann::json& j, 
-    rcbe::math::Matrix<rcbe::core::EngineScalar, 3, 3>& value
-)
+void adl_serializer<rcbe::math::Matrix3x3>:: from_json(const nlohmann::json& j, rcbe::math::Matrix3x3& value)
+{
+    auto data = j.at("matrix").get<rcbe::math::Matrix3x3::storage_type>();
+    auto rows = j.at("rows").get<rcbe::core::EngineIntergral>();
+    auto cols = j.at("cols").get<rcbe::core::EngineIntergral>();
+    assert(rows == 3 && cols == 3);
```
Comment (radameon) [inline engine_datamodel/src/math/Matrix.cpp:22]: Replace with `throw`

File: engine_datamodel/src/math/Vector.cpp [hunk 9]
```diff
@@ -19,6 +20,31 @@ std::istream &operator>>(std::istream &is, rcbe::math::Vector3d &vec)
...
+}
+
+rcbe::math::Vector3d operator*(const rcbe::core::EngineScalar &lhs, const rcbe::math::Vector3d &rhs)
+{
+    return rcbe::math::Vector3d { rhs.x() * lhs, rhs.y() * lhs, rhs.z() * lhs };
+}
+
+rcbe::math::Vector3d operator*(const rcbe::math::Vector3d &lhs, const rcbe::core::EngineScalar &rhs)
+{
+    return rcbe::math::Vector3d { lhs.x() * rhs, lhs.y() * rhs, lhs.z() * rhs };
```
Comment (radameon) [inline engine_datamodel/src/math/Vector.cpp:45]: Call the above opertaor

File: engine_datamodel/include/data_types/config/data_model_config.hpp [hunk 10]
```diff
@@ -1,6 +1,6 @@
 #ifndef RCBE_DATA_MODEL_CONFIG_HPP
 #define RCBE_DATA_MODEL_CONFIG_HPP
 
-#include <common/fundamentals/types.hpp>
+#include "engine_common/include/common/fundamentals/types.hpp"
```
[the remark this answers is not in the document]
Comment (Sckorn) [reply engine_datamodel/include/data_types/config/data_model_config.hpp]: Changed

File: engine_datamodel/include/data_types/core/CObject.hpp [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+        _value { o }
+        , _id { std::hash<size_t>()(std::chrono::steady_clock::now()) }
+        {}
+
+        size_t hash() const override
+        {
+            return std::hash<size_t>()(_id);
+        }
+
+        size_t id() const
```
[the remark this answers is not in the document]
Comment (Sckorn) [reply engine_datamodel/include/data_types/core/CObject.hpp]: Fixed

File: engine_datamodel/include/data_types/geometry/Mesh.hpp [hunk 12]
```diff
@@ -90,6 +93,9 @@ class Mesh
 
     const color_type &color() const;
 
+    const transform_type transform() const;
+    void set_transform(const transform_type &t);
```
[the remark this answers is not in the document]
Comment (Sckorn) [reply engine_datamodel/include/data_types/geometry/Mesh.hpp:98]: Done

File: engine_datamodel/include/data_types/math/Matrix.hpp [hunk 13]
```diff
@@ -32,30 +81,95 @@ class Matrix
...
+        for (size_t i = 0; i < DimRow; ++i)
+        {
+            _tmp[i] = _m[i * DimCol + index];
+        }
+
+        return column_type {_tmp};
+    }
+
+    // row-major
+    const value_type &at(size_t row, size_t col) const
```
[the remark this answers is not in the document]
Comment (Sckorn) [reply engine_datamodel/include/data_types/math/Matrix.hpp:132]: Renge check implmented inside nested `at` method, so additional check is redundant. Also `operator[]` would be restricted to only one index, which makes this operator's work mor complicated.

File: engine_datamodel/include/data_types/math/Quaternion.hpp [hunk 14]
```diff
@@ -0,0 +1,176 @@
...
+#define RCBE_QUATERNION_HPP
+
+#include <cmath>
+
+#include <data_types/math/Vector.hpp>
+
+namespace rcbe::math
+{
+template <typename ValueType>
+class Quaternion // TODO: implement conjuction and inverse
```
[the remark this answers is not in the document]
Comment (Sckorn) [reply engine_datamodel/include/data_types/math/Quaternion.hpp]: Done

File: engine_datamodel/include/data_types/math/Quaternion.hpp [hunk 15]
```diff
@@ -0,0 +1,176 @@
...
+    using storage_type = Vector<value_type, dimension>;
+
+    Quaternion(const Vector3d &v, const value_type s)
+    :
+    q_ { v.x(), v.y(), v.z(), s }
+    {}
+
+    Quaternion(const value_type yaw, const value_type pitch, const value_type roll)
+    {
+        const auto deg_to_rad = [](const value_type angle) {
```
[the remark this answers is not in the document]
Comment (Sckorn) [reply engine_datamodel/include/data_types/math/Quaternion.hpp]: Done

File: engine_datamodel/src/geometry/Mesh.cpp [hunk 16]
```diff
@@ -96,6 +96,26 @@ const Mesh::color_type &Mesh::color() const
...
+{
+    _transform = t;// * _transform;
```
[the remark this answers is not in the document]
Comment (Sckorn) [reply engine_datamodel/src/geometry/Mesh.cpp]: Fixed

File: engine_datamodel/include/data_types/math/Quaternion.hpp [hunk 17]
```diff
@@ -0,0 +1,176 @@
...
+        vector_type tmp {conj.x(), conj.y(), conj.z()};
+        auto vec = tmp * norm_val;
+
+        return {vec, scalar};
+    }
+
+private:
+    storage_type q_;
+};
+}
```
[the remark this answers is not in the document]
Comment (Sckorn) [reply engine_datamodel/include/data_types/math/Quaternion.hpp:184]: Done

File: engine_datamodel/src/geometry/Mesh.cpp (hunk 6 above)
Comment (Sckorn) [reply engine_datamodel/src/geometry/Mesh.cpp]: Done

File: engine_datamodel/src/geometry/Mesh.cpp (hunk 7 above)
Comment (Sckorn) [reply engine_datamodel/src/geometry/Mesh.cpp]: Done

File: engine_datamodel/src/math/Vector.cpp (hunk 9 above)
Comment (Sckorn) [reply engine_datamodel/src/math/Vector.cpp]: Done

File: engine_datamodel/src/math/Matrix.cpp (hunk 8 above)
Comment (Sckorn) [reply engine_datamodel/src/math/Matrix.cpp]: Done

File: engine_datamodel/include/data_types/math/Transform.hpp (hunk 5 above)
Comment (Sckorn) [reply engine_datamodel/include/data_types/math/Transform.hpp:27]: Done

File: engine_datamodel/include/data_types/math/Matrix.hpp (hunk 1 above)
Comment (Sckorn) [reply engine_datamodel/include/data_types/math/Matrix.hpp]: Done

File: engine_datamodel/include/data_types/math/Matrix.hpp (hunk 2 above)
Comment (Sckorn) [reply engine_datamodel/include/data_types/math/Matrix.hpp]: Done

File: engine_datamodel/include/data_types/math/Quaternion.hpp (hunk 3 above)
Comment (Sckorn) [reply engine_datamodel/include/data_types/math/Quaternion.hpp]: Done

File: engine_datamodel/include/data_types/math/Quaternion.hpp (hunk 4 above)
Comment (Sckorn) [reply engine_datamodel/include/data_types/math/Quaternion.hpp]: Done

File: engine_datamodel/include/data_types/math/Quaternion.hpp [hunk 18]
```diff
@@ -0,0 +1,186 @@
...
+    {
+        auto copy = *this;
+
+        copy.normalize();
+        return copy;
+    }
+
+    friend Quaternion operator+(const Quaternion &q1, const Quaternion &q2)
+    {
+        return { { q1.q_.at(0) + q2.q_.at(0), q1.q_.at(1) + q2.q_.at(1), q1.q_.at(2) + q2.q_.at(2) }, q1.q_.at(3) + q2.q_.at(3) };
```
Comment (radameon) [inline engine_datamodel/include/data_types/math/Quaternion.hpp:140]: Use `getVectorPart` and `getScalarPart`

File: engine_datamodel/include/data_types/math/Vector.hpp [hunk 19]
```diff
@@ -61,6 +66,14 @@ class Vector
         return _v[index];
     }
 
+    const value_type &at(const size_t index) const
+    {
+        if (index < 0 || index >= dimension)
+            throw std::out_of_range(index + " is out of range.");
+
+        return _v[index];
```
Comment (radameon) [inline engine_datamodel/include/data_types/math/Vector.hpp:74]: Use `_v.at()`

File: engine_datamodel/include/data_types/math/Quaternion.hpp (hunk 18 above)
Comment (Sckorn) [reply engine_datamodel/include/data_types/math/Quaternion.hpp]: Done

File: engine_datamodel/include/data_types/math/Vector.hpp (hunk 19 above)
Comment (Sckorn) [reply engine_datamodel/include/data_types/math/Vector.hpp]: Done

**Pull request merged**